File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root: return []
        final = []
        def dfs(node):
            final.append(node.val)
            if node.left: dfs(node.left)
            if node.right: dfs(node.right)
        dfs(root)
        return final

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */


void dfs(struct TreeNode* node, int* arr, int* returnSize){
    arr[(*returnSize)++] = node->val; // increment counter after writing to final
    if(node->left != NULL){
        dfs(node->left, arr, returnSize);
    }
    if(node->right != NULL){
        dfs(node->right, arr, returnSize);
    }
}   

int* preorderTraversal(struct TreeNode* root, int* returnSize){
    int* final = (int*)malloc(100 * sizeof(int));
    *returnSize = 0;
    if(root == NULL){
        return final;
    }
    dfs(root, final, returnSize);
    for(int i = 0; i < *returnSize; i ++){
        printf("%d, ", final[i]);
    }
    return final;
}   